File: mf/python/example_poll_impedance.py
# Copyright 2018 Zurich Instruments AG

"""
Zurich Instruments LabOne Python API Example

Demonstrate how to obtain impedance data using ziDAQServer's blocking
(synchronous) poll() command.
Connect to the device specified by device_id and obtain
impedance data using ziDAQServer's blocking (synchronous) poll() command.

Requirements:
    * LabOne Version >= 20.02
    * Instruments:
        1 x Instrument with IA option
    * Connect Impedance Testfixture and attach a 1kOhm resistor

Usage:
    example_autoranging_impedance.py [options] <device_id>
    example_autoranging_impedance.py -h | --help

Arguments:
    <device_id>  The ID of the device [device_type: MF*(IA)]

Options:
    -h --help              Show this screen.
    -s --server_host IP    Hostname or IP address of the dataserver [default: localhost]
    -p --server_port PORT  Port number of the data server [default: 8004]
    --no-plot              Hide plot of the recorded data.

Returns:
    sample  The impedance sample dictionary as returned by poll.

Raises:
    Exception     If the specified devices do not match the requirements.
    RuntimeError  If the devices is not "discoverable" from the API.

See the LabOne Programming Manual for further help:
https://docs.zhinst.com/labone_programming_manual/
"""
import time
import zhinst.utils
import matplotlib.pyplot as plt
import numpy as np

import example_autoranging_impedance


def run_example(
    device_id: str,
    server_host: str = "localhost",
    server_port: int = 8004,
    plot: bool = True,
):
    """run the example."""

    apilevel_example = 6  # The API level supported by this example.
    # Call a zhinst utility function that returns:
    # - an API session `daq` in order to communicate with devices via the data server.
    # - the device ID string that specifies the device branch in the server's node hierarchy.
    # - the device's discovery properties.
    (daq, device, _) = zhinst.utils.create_api_session(
        device_id, apilevel_example, server_host=server_host, server_port=server_port
    )
    zhinst.utils.api_server_version_check(daq)

    # Create a base configuration: Disable all available outputs, awgs, demods, scopes,...
    zhinst.utils.disable_everything(daq, device)

    # We use the auto-range example to perform some basic device configuration
    # and wait until signal input ranges have been configured by the device.
    example_autoranging_impedance.run_example(device)

    # Subscribe to the impedance sample node path.
    imp_index = 0
    path = "/%s/imps/%d/sample" % (device, imp_index)
    daq.subscribe(path)

    # Sleep for demonstration purposes: Allow data to accumulate in the data
    # server's buffers for one second: poll() will not only return the data
    # accumulated during the specified poll_length, but also for data
    # accumulated since the subscribe() or the previous poll.
    sleep_length = 1.0
    # For demonstration only: We could, for example, be processing the data
    # returned from a previous poll().
    time.sleep(sleep_length)

    # Poll the subscribed data from the data server. Poll will block and record
    # for poll_length seconds.
    poll_length = 0.1  # [s]
    poll_timeout = 500  # [ms]
    poll_flags = 0
    poll_return_flat_dict = True
    data = daq.poll(poll_length, poll_timeout, poll_flags, poll_return_flat_dict)

    # Unsubscribe from all paths.
    daq.unsubscribe("*")

    # Check the dictionary returned is non-empty
    assert (
        data
    ), "poll() returned an empty data dictionary, did you subscribe to any paths?"

    # The data returned is a dictionary of dictionaries that reflects the node's path.
    # Note, the data could be empty if no data had arrived, e.g., if the imps
    # were disabled or had transfer rate 0.
    assert path in data, "The data dictionary returned by poll has no key `%s`." % path

    # Access the impedance sample using the node's path. For more information
    # see the data structure documentation for ZIImpedanceSample in the LabOne
    # Programming Manual.
    impedance_sample = data[path]

    # Get the sampling rate of the device's ADCs, the device clockbase in order
    # to convert the sample's timestamps to seconds.
    clockbase = float(daq.getInt("/%s/clockbase" % device))

    dt_seconds = (
        impedance_sample["timestamp"][-1] - impedance_sample["timestamp"][0]
    ) / clockbase
    num_samples = len(impedance_sample["timestamp"])
    print(
        f"poll() returned {num_samples} samples of impedance data spanning {dt_seconds:.3f} \
            seconds."
    )
    print(f"Average measured resitance: {np.mean(impedance_sample['param0'])} Ohm.")
    print(f"Average measured capacitance: {np.mean(impedance_sample['param1'])} F.")

    if plot:

        # Convert timestamps from ticks to seconds via clockbase.
        t = (
            impedance_sample["timestamp"] - impedance_sample["timestamp"][0]
        ) / clockbase

        plt.close("all")
        # Create plot
        _, axes = plt.subplots(2, sharex=True)
        axes[0].plot(t, impedance_sample["param0"])
        axes[0].set_title("Impedance Parameters")
        axes[0].grid(True)
        axes[0].set_ylabel(r"Resistance ($\Omega$)")
        axes[0].autoscale(enable=True, axis="x", tight=True)

        axes[1].plot(t, impedance_sample["param1"])
        axes[1].grid(True)
        axes[1].set_ylabel(r"Capacitance (F)")
        axes[1].set_xlabel("Time (s)")
        axes[1].autoscale(enable=True, axis="x", tight=True)

        plt.draw()
        plt.show()

    return data


if __name__ == "__main__":
    import sys
    from pathlib import Path

    cli_util_path = Path(__file__).resolve().parent / "../../utils/python"
    sys.path.insert(0, str(cli_util_path))
    cli_utils = __import__("cli_utils")
    cli_utils.run_commandline(run_example, __doc__)
    sys.path.remove(str(cli_util_path))

File: mf/matlab/example_poll_impedance.m
function data = example_poll_impedance(device_id, varargin)
% EXAMPLE_AUTORANGING_IMPEDANCE Demonstrate how to obtain impedance
% data using ziDAQServer's blocking (synchronous) poll() command.
%
% USAGE DATA = EXAMPLE_POLL_IMPEDANCE(DEVICE_ID)
%
% The example demonstrates how to poll impedance data.
% DEVICE_ID should be a string, e.g., 'dev1000'.
%
% NOTE Additional configuration: Connect Impedance Testfixture and attach a 1kOhm
% resistor
%
% NOTE Please ensure that the ziDAQ folders 'Driver' and 'Utils' are in your
% Matlab path. To do this (temporarily) for one Matlab session please navigate
% to the ziDAQ base folder containing the 'Driver', 'Examples' and 'Utils'
% subfolders and run the Matlab function ziAddPath().
% >>> ziAddPath;
%
% Use either of the commands:
% >>> help ziDAQ
% >>> doc ziDAQ
% in the Matlab command window to obtain help on all available ziDAQ commands.
%
% Copyright 2008-2018 Zurich Instruments AG

clear ziDAQ;

if ~exist('device_id', 'var')
    error(['No value for device_id specified. The first argument to the ' ...
           'example should be the device ID on which to run the example, ' ...
           'e.g. ''dev1000''.'])
end

% Check the ziDAQ MEX (DLL) and Utility functions can be found in Matlab's path.
if ~(exist('ziDAQ') == 3) && ~(exist('ziCreateAPISession', 'file') == 2)
    fprintf('Failed to either find the ziDAQ mex file or ziDevices() utility.\n')
    fprintf('Please configure your path using the ziDAQ function ziAddPath().\n')
    fprintf('This can be found in the API subfolder of your LabOne installation.\n');
    fprintf('On Windows this is typically:\n');
    fprintf('C:\\Program Files\\Zurich Instruments\\LabOne\\API\\MATLAB2012\\\n');
    return
end

% The API level supported by this example.
apilevel_example = 6;
% Create an API session; connect to the correct Data Server for the device.
[device, props] = ziCreateAPISession(device_id, apilevel_example);
ziApiServerVersionCheck();

% This example requires an MF device with IA option.
if ~ismember('IA', props.options)
    installed_options_str = ''; % Note: strjoin() unavailable in Matlab 2009.
    for option=props.options
      installed_options_str = [installed_options_str, ' ', option{1}];
    end
    error(['Required option set not satisfied. This example requires an MF ' ...
           'Instrument with the IA Option installed. Device `%s` reports ' ...
           'devtype `%s` and options `%s`.'], ...
          device, props.devicetype, installed_options_str);
end

% We use the auto-range example to perform some basic device configuration
% and wait until signal input ranges have been configured by the device.
example_autoranging_impedance(device);

% The impedance channel to use.
imp_c = '0';
imp_index = 1;

% Subscribe to the impedance sample
% The Data Server's node tree uses 0-based indexing; Matlab uses
% 1-based indexing:
imp_node_path = ['/' device, '/imps/' imp_c '/sample'];
ziDAQ('subscribe', imp_node_path);

% Sleep for demonstration purposes: Allow data to accumulate in the data
% server's buffers for one second: poll() will not only return the data
% accumulated during the specified poll_length, but also for data
% accumulated since the subscribe() or the previous poll.
pause(1.0);

poll_duration = 0.02;
poll_timeout = 100;
data = ziDAQ('poll', poll_duration, poll_timeout);

% Unsubscribe from all paths.
ziDAQ('unsubscribe', '*');

if ziCheckPathInData(data, imp_node_path)
    sample = data.(device).imps(imp_index).sample;

    % Get the sampling rate of the device's ADCs, the device clockbase in order to convert the sample's timestamps to
    % seconds.
    clockbase = double(ziDAQ('getInt', ['/' device '/clockbase']));
    t0 = double(sample.timestamp(1));
    dt_seconds = (double(sample.timestamp(end)) - t0)/clockbase;
    num_samples= length(sample.timestamp);

    fprintf('Poll returned %d samples of impedance data spanning %.3f s.\n', num_samples, dt_seconds)
    fprintf('Average measured resistance: %.3e Ohm.\n', mean(sample.param0))
    fprintf('Average measured capacitance: %.3e F.\n', mean(sample.param1))

    t = (double(sample.timestamp) - t0)/clockbase;

    figure(1)
    clf;
    title('\\bImpedance Parameter Data');
    hold on;
    ax1 = subplot(2, 1, 1);
    plot(t, sample.param0);
    grid on;
    ylabel('Resistance [Ohm]')
    ax2 = subplot(2, 1, 2);
    plot(t, sample.param1);
    grid on;
    ylabel('Capcitance [F]')
    xlabel('Time (s)')
    x = xlim(ax1);
    xlim(ax1, [min(x) max(x)]);
    linkaxes([ax1, ax2], 'x')
else
    error(['Poll did not return the expected data (', imp_node_path, ') - data transfer rate is 0?']);
end
